taichi/backends/codegen_metal.cpp

```diff
@@ -33,6 +35,345 @@ class MetalKernelCodegen : public IRVisitor {
+    } else {
+      emit("const {} {} = static_cast<{}>({});",
+           metal_data_type_name(stmt->element_type()), stmt->raw_name(),
+           metal_data_type_name(stmt->cast_type), stmt->operand->raw_name());
+    }
+  }
+
+  void visit(BinaryOpStmt *bin) override {
+    const auto dt_name = metal_data_type_name(bin->element_type());
+    const auto lhs_name = bin->lhs->raw_name();
+    const auto rhs_name = bin->rhs->raw_name();
+    const auto bin_name = bin->raw_name();
+    if (bin->op_type == BinaryOpType::floordiv) {
+      if (is_integral(bin->element_type())) {
+        const auto intm = fmt::format("{}_intermediate_", bin_name);
+        emit("const {} {} = ({} / {});", dt_name, intm, lhs_name, rhs_name);
+        // Should we construct an AST for this?
+        const auto expr_str = fmt::format(
+            "(({lhs} * {rhs} < 0) && ({rhs} * {intm} != {lhs})) ? ({intm} - 1) "
+            ": {intm}",
+            fmt::arg("lhs", lhs_name), fmt::arg("rhs", rhs_name),
+            fmt::arg("intm", intm));
+        emit("const {} {} = ({});", dt_name, bin_name, expr_str);
+      } else {
+        emit("const {} {} = floor({} / {});", dt_name, bin_name, lhs_name,
+             rhs_name);
+      }
+      return;
+    }
+    const auto binop = metal_binary_op_type_symbol(bin->op_type);
+    if (is_metal_binary_op_infix(bin->op_type)) {
+      emit("const {} {} = ({} {} {});", dt_name, bin_name, lhs_name, binop,
+           rhs_name);
+    } else {
+      // This is a function call
+      emit("const {} {} =  {}({}, {});", dt_name, bin_name, binop, lhs_name,
+           rhs_name);
+    }
+  }
+
+  void visit(TernaryOpStmt *tri) override {
+    TC_ASSERT(tri->op_type == TernaryOpType::select);
+    emit("const {} {} = ({}) ? ({}) : ({});",
+         metal_data_type_name(tri->element_type()), tri->raw_name(),
+         tri->op1->raw_name(), tri->op2->raw_name(), tri->op3->raw_name());
+  }
+
+  void visit(AtomicOpStmt *stmt) override {
+    TC_ASSERT(stmt->width() == 1);
+    TC_ASSERT(stmt->op_type == AtomicOpType::add);
+    const auto dt = stmt->val->element_type();
+    if (dt == DataType::i32) {
+      emit(
+          "const auto {} = atomic_fetch_add_explicit((device atomic_int*){}, "
+          "{}, "
+          "metal::memory_order_relaxed);",
+          stmt->raw_name(), stmt->dest->raw_name(), stmt->val->raw_name());
+    } else if (dt == DataType::f32) {
+      // A huge hack! Metal does not support atomic floating point numbers
+      // natively.
+      const auto dest_name = stmt->dest->raw_name();
+      const auto cas_ok = fmt::format("{}_cas_ok_", dest_name);
+      const auto old_val = fmt::format("{}_old_", dest_name);
+      const auto new_val = fmt::format("{}_new_", dest_name);
+      emit("bool {} = false;", cas_ok);
+      emit("float {} = 0.0f;", stmt->raw_name());
+      emit("while (!{}) {{", cas_ok);
+      push_indent();
+      emit("float {} = *{};", old_val, dest_name);
```

Comment:
If we have more stuff like this in the future, maybe we can create a "metal_runtime.h" with something like `f32 atomic_add(f32* dst, f32 val)` to be included for each kernel. Then we can simply call the functions here.

Reply:
Good point, I will do that in a following PR!